### PR TITLE
New Language: V

### PR DESCRIPTION
**Pull request by Spydr06**
V (`vlang`) is a simple, fast and safe general purpose programming language compiling to human-readable C.

#### Key features of V:

- simplicity, "only one way to do things"
- performance as fast as C
- safety: no null, no globals, no undefined behavior, immutability
- automatic C to V translation, good C interoperability
- hot code reloading
- flexible memory management (GC by default, manual, arena allocation, autofree)
- other compilation backends like JavaScript, `native` (wip), `wasm` (wip) or interpreted

#### Links:

Source code: https://github.com/vlang/v
Official website: https://vlang.io
Web playground: https://play.vlang.io/

infra PR: https://github.com/compiler-explorer/infra/pull/1058

#### Things this PR adds:

- General support for V
- Code inspection for the C, Go and JavaScript backends
- Support for the v formatter `v fmt`
- Some V example code

#### Things not implemented:

- Support for binary-output backends like `native` and `wasm`
- Support for running programs

### CODE REVIEW

File: lib/languages.ts [hunk 1]
```diff
@@ -648,6 +648,17 @@ const definitions: Record<LanguageKey, LanguageDefinition> = {
         previewFilter: null,
         monacoDisassembly: null,
     },
+    v: {
+        name: 'V',
+        monaco: 'v',
+        extensions: ['.v'],
```
Comment (Delta456) [inline lib/languages.ts:654]: You might also want to add `.vsh`

File: static/modes/v-mode.ts [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+            'struct',
+            'true',
+            'type',
+            'typeof',
+            'union',
+            'unsafe',
+            'volatile',
+        ],
+
+        typeKeywords: [
```
Comment (Delta456) [inline static/modes/v-mode.ts:84]: `usize` and `isize` needs to be added too.

Comment (Spydr06) [reply static/modes/v-mode.ts:84]: fixed in [627bdfb](https://github.com/compiler-explorer/compiler-explorer/pull/5297/commits/627bdfb5f0377a4e6164f791f640ac8568be2aed)

File: lib/languages.ts (hunk 1 above)
Comment (Spydr06) [reply lib/languages.ts]: fixed in [627bdfb](https://github.com/compiler-explorer/compiler-explorer/pull/5297/commits/627bdfb5f0377a4e6164f791f640ac8568be2aed)

File: static/modes/v-mode.ts [hunk 3]
```diff
@@ -0,0 +1,252 @@
+// Copyright (c) 2012, Compiler Explorer Authors
```
Comment (partouf) [inline static/modes/v-mode.ts:1]: wrong date

File: lib/compilers/v.ts [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+export class VCompiler extends BaseCompiler {
+    outputFileExt = `.${V_DEFAULT_BACKEND}`;
+
+    static get key() {
+        return 'v';
+    }
+
+    constructor(info: PreliminaryCompilerInfo, env) {
+        super(info, env);
+        this.compiler.supportsBinary = false;
```
Comment (partouf) [inline lib/compilers/v.ts:43]: might be cleaner if you just set supportsBinary and supportsExecute to false in the vala.defaults.properties file

Comment (partouf) [reply lib/compilers/v.ts:43]: and the amazon.properties

File: etc/config/v.amazon.properties [hunk 5]
```diff
@@ -0,0 +1,18 @@
+# Disabled: v v04
```
Comment (partouf) [inline etc/config/v.amazon.properties:1]: this disabled comment should not be needed

**Review by partouf — CHANGES_REQUESTED** [review]
Thanks for the PR!
Looks good overall, added a couple of tiny comments

File: etc/config/v.amazon.properties (hunk 5 above)
Comment (Spydr06) [reply etc/config/v.amazon.properties:1]: Without this comment, the pre-commit hook fails:
image

File: static/modes/v-mode.ts (hunk 3 above)
Comment (Spydr06) [reply static/modes/v-mode.ts]: fixed in [49c6438](https://github.com/compiler-explorer/compiler-explorer/pull/5297/commits/49c6438c7ca7932b4acc7a0ab5b35bec117d78e0)

File: lib/compilers/v.ts (hunk 4 above)
Comment (Spydr06) [reply lib/compilers/v.ts]: fixed in [49c6438](https://github.com/compiler-explorer/compiler-explorer/pull/5297/commits/49c6438c7ca7932b4acc7a0ab5b35bec117d78e0)

**Comment by Spydr06** [comment]
Thanks for the review! Most things are fixed now

File: etc/config/v.amazon.properties (hunk 5 above)
Comment (partouf) [reply etc/config/v.amazon.properties:1]: on line 6 you have `defaultCompiler=v`, this should be `defaultCompiler=v04` and on line 12 it should be `group.v.compilers=v04` - that should fix these errors

Comment (Spydr06) [reply etc/config/v.amazon.properties]: awesome, fixed!

**Review by partouf — APPROVED** [review]

**Comment by mattgodbolt** [comment]
This will go live, but won't work: in testing:

```
V panic: folder: /opt/compiler-explorer/v-2023.30/.vmodules/cache/, error: Permission denied
```

Happy to help debug this, but I'm not going to revert; we'll just have `V` broken on the site for a bit :)

**Comment by Spydr06** [comment]
Thanks for the merge.

V needs write permission to a `.vmodules` folder.
I chose this location over in the infra repo, since I thought the V install location was writable...
What would be a better location for this folder?

**Comment by Spydr06** [comment]
https://github.com/compiler-explorer/compiler-explorer/pull/5385

**Comment by mattgodbolt** [comment]
> It doesn't have to be temporary ...

but if it wasn't it could reasonably be used to leak information about other people's compilations and/or would interact badly with two or more simultanous compilations at once :)

Thanks for the PR! Will try and look tonight!

**Comment by joe-conigliaro** [comment]
Thanks @Spydr06, nice work 🥳

**Pull request merged**